File: pset6/credit/credit.py
from cs50 import get_int

# user input for integer greater than 0
while True:
    ccNumber = get_int("Number: ")
    if ccNumber >= 0:
        break

# count number of digits
ccNumberString = str(ccNumber)
numDigits = len(ccNumberString)

# to maximize efficiency, let's just check up front if the number of digits is a valid length for a ccNumber.
if numDigits not in(13, 15, 16):
    print("INVALID")
    exit()

# make a duplicate of ccNumber to iterate through
ccNumberIterator = ccNumber
# strip off the first digit of the iterator
ccNumberIterator //= 10
# if ccNumber was 12345, it should be stripped to 1234


firstDigitSetSum = 0
secondDigitSetSum = 0
currentDigit = 0

# for the first set of digits, multiply by two and add the sum of the products' digits together.
for digit in range(numDigits - 1):
    # `numDigits - 1` because we stripped off the last digit
    currentDigit = ccNumberIterator % 10
    ccNumberIterator //= 100
    currentDigit *= 2

    # add the digits of the current product. There will be a maximum of two digits, so the following works:
    currentDigit = (currentDigit // 10) + (currentDigit % 10)
    firstDigitSetSum += currentDigit

    # if we have exhausted all digits from first digit set
    if ccNumberIterator == 0:
        break

# rest ccNumberIterator and currentDigit
ccNumberIterator = ccNumber
currentDigit = 0

# for the second set of digits
for digit in range(numDigits):
    currentDigit = ccNumberIterator % 10
    ccNumberIterator //= 100
    secondDigitSetSum += currentDigit
    if ccNumberIterator == 0:
        break

finalSum = firstDigitSetSum + secondDigitSetSum

firstTwoDigits = int(str(ccNumber)[:2])

# if final digit of finalSum != 0, INVALID
if finalSum % 10 != 0:
    print("INVALID")
    exit()

# AMEX (15 digits, start w 34 or 37)
if numDigits == 15:
    if firstTwoDigits == 34 or firstTwoDigits == 37:
        print("AMEX")
        exit()
    else:
        print("INVALID")
        exit()
# MasterCard (16 digits, start with 51, 52, 53, 54, or 55)
elif numDigits == 16:
    if firstTwoDigits in(51, 52, 53, 54, 55):
        print("MASTERCARD")
        exit()
    elif firstTwoDigits // 10 == 4:
        print("VISA")
        exit()
# Visa (13 and 16 digit numbers, start with 4)
elif numDigits == 13 and firstTwoDigits // 10 == 4:
    print("MASTERCARD")
    exit()

File: pset1/credit.c
// This program determines whether a provided credit card number (user-input) is valid according to Luhn's algorithm.
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    // user-input for number greater than 0
    long long ccNumber;
    do
    {
        ccNumber = get_long_long("Number: ");
    }
    while (ccNumber <= 0);

    // count number of digits
    int numDigits = 0;
    long long digitValidator = ccNumber;

    while (digitValidator != 0)
    {
        // divides by 10 until the long long is == 0.. i.e: 400/10 -> 40; 40/10 -> 4; 4/10 -> 0
        // this digit counter works by counting the number of times it takes to divide by 10 to get to 0.
        digitValidator = digitValidator / 10;
        numDigits++;
    }

    // first checksum on the first set of digits
    long long firstDigitSet = ccNumber;
    int digits = 2 * ((firstDigitSet / 10) % 10);
    firstDigitSet /= 10;
    int sum = (digits / 10) + (digits % 10);

    for (int i = 0; i < numDigits; i++)
    {
        // isolating the necessary digits and multiplying by 2
        digits = 2 * ((firstDigitSet / 100) % 10);
        firstDigitSet /= 100;
        // adding the products' digits
        sum += (digits / 10) + (digits % 10);
    }

    // second checksum on second set of digits
    long long secondDigitSet = ccNumber;
    sum += secondDigitSet % 10;
    for (int i = 0; i < numDigits; i++)
    {
        // isolating the other digits
        digits = ((secondDigitSet / 100) % 10);
        sum += (digits / 10) + (digits % 10);
        secondDigitSet /= 100;
    }

    // Isolating first two digits for use in final check
    int firstTwoDigits = 0;

    // Checking for digit length and company specs
    if (numDigits == 15 && sum % 10 == 0)
    {
        firstTwoDigits = ccNumber / 10000000000000;
        switch (firstTwoDigits)
        {
            case 34:
            case 37:
                printf("AMEX\n");
                break;

            default:
                printf("INVALID\n");
                break;
        }
    }
    else if (numDigits == 13 && sum % 10 == 0)
    {
        firstTwoDigits = ccNumber / 100000000000;
        if ((firstTwoDigits / 10) % 10 == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else if (numDigits == 16 && sum % 10 == 0)
    {
        firstTwoDigits = ccNumber / 100000000000000;
        if ((firstTwoDigits / 10) % 10 == 4)
        {
            printf("VISA\n");
        }
        else if (firstTwoDigits == 51 || firstTwoDigits == 52 || firstTwoDigits == 53 || firstTwoDigits == 54
                 || firstTwoDigits == 55)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}